activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,160 @@
+
+module ActiveSupport
+  module Cache
+    # This class is responsible for cohercing input into a valid
+    # key format and cache version for cache stores.
+    #
+    # The goal of this class is to provide a composable way
+    # to build keys. You can put an instance of Key
+    # into another instance of Key and it should produce
+    # the same `cache_key` and `cache_version`.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k2 = Key.new(k1)
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #
+    # In addition to passing a key into the initialize method
+    # key fragments can be added onto an existing key through the
+    # `<<` method.
+    #
+    # Example:
+    #
+    #   key = Key.new
+    #   key << "foo"
+    #   key << "bar"
+    #   key.cache_key
+    #   # => "foo/bar"
+    #
+    # Multiple key entries generate the same key regardless of how
+    # they are added. I.e. if they're added via multiple arrays
+    # or directly via the `<<` operator.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k1 << "stay-puft"
+    #   k2 = Key.new(["foo", "bar"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = Key.new(["foo", ["bar"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.to_key(key)
+        return key if key.is_a?(Key)
```

Comment:
Do you happen to have a feel for which scenario (passing in a `Key` vs something else) is more common? I did a microbenchmark (admittedly with strings/arrays) and a ternary would be faster for the else condition.

```
one - the way you have now
two - ternary
Warming up --------------------------------------
            one - if   184.196k i/100ms
            two - if   157.414k i/100ms
          one - else   172.947k i/100ms
          two - else   225.391k i/100ms
Calculating -------------------------------------
            one - if      5.276M (± 3.9%) i/s -     26.340M in   5.000433s
            two - if      3.418M (± 5.6%) i/s -     17.158M in   5.037682s
          one - else      3.710M (± 9.9%) i/s -     18.332M in   5.004147s
          two - else      6.629M (± 5.8%) i/s -     33.132M in   5.018285s

Comparison:
          two - else:  6629185.6 i/s
            one - if:  5275983.5 i/s - 1.26x  slower
          one - else:  3709665.6 i/s - 1.79x  slower
            two - if:  3417664.4 i/s - 1.94x  slower
```

Reply:
I'm not able to reproduce that. Can you please share your bench script when posting results. Here's mine https://gist.github.com/schneems/8412f1c85672c76fc9946111138d9e66.

> Do you happen to have a feel for which scenario (passing in a Key vs something else) is more common?

Ideally, we want to introduce a key object as soon as possible otherwise if a non `Key` object such as an array makes it all the way down to `normalize_key` then duplicate work will end up happening. That's why there are several cases where we're ensuring the key we are working with is a `Key` object.

Reply:
https://gist.github.com/oniofchaos/297eaaad5ec3d0b4dde394797bcb34ed

Reply:
Whoops....at any rate, looks like this code got refactored. Thanks.